Fix UseNullPropagation codefix doesn't recognize `x is null ? null : x.y` pattern

schovanec opened the pull request:
Fixes #23043. This should handle both the `x is null` and the `ReferenceEquals(x, null)` scenarios.

### Customer scenario

Visual Studio 2017 contains a "Use Null Propagation" analyzer and code fix that helps users convert conditional code (`if` statements and ternary expressions) into the new `?.` syntax. While the analysis already supports the most common styles of null checks, there were two specific cases that were simply ignored:

* Conditions based on the new pattern syntax: `x is null`
* Conditions based on calls to `object.ReferenceEquals`

### Bugs this fixes

Fixes #23043 

### Workarounds, if any

None needed

### Risk

Low¹. Changes are restricted to a single analyzer/fix, and tests² are added for all newly supported cases.

¹ ✔️ I need to apply a Fix All on Roslyn before and after this change to review the difference for bugs. (Confirmed via https://github.com/dotnet/roslyn/pull/23610#issuecomment-350328171)
² ✔️ Negative tests would be good as well. I am requesting them in review. (Confirmed via https://github.com/dotnet/roslyn/pull/23610#issuecomment-350294478)

### Performance impact

Should be low. The code was reviewed with consideration for known performance sensitivities in analyzers, and will be run through AnalyzerRunner prior to final release.

### Is this a regression from a previous update?

No.

### Root cause analysis

Previously identified optional work from a prior feature.

### How was the bug found?

Spec review of the original feature.

### Test documentation updated?

No.

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=23610) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=23610) before we can accept your contribution.<br/><br/>:x: schovanec [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=23610)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn?pullRequest=23610) it.</sub>

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/UseNullPropagation/UseNullPropagationTests.cs:
```diff
@@ -478,5 +478,125 @@ void Main(S? s)
+    public int? f;
+    void M(C c)
+    {
+        int? x = [||]ReferenceEquals(c, null) ? null : c.f;
```
Cool changes :)

Consider also supporting:

```!ReferenceEquals(c, null) ? c.f : null```

schovanec replied:
That’s a good idea. I’ll look into that later tonight.

sharwell replied:
:bulb: Also have an opportunity here for `!(c is null)`

schovanec replied:
It was quite simple to implement this so that logical negation would be recognized on any condition, including `!(c == null)` and `!(c != null)`. If to only support it on the `ReferenceEquals` and pattern syntax I can change it.

CyrusNajmabadi commented on src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:
```diff
@@ -49,18 +50,27 @@ protected AbstractUseNullPropagationDiagnosticAnalyzer()
+
+                var objectType = startContext.Compilation.GetSpecialType(SpecialType.System_Object);
+                var referenceEqualsMethod = objectType?.GetMembers(nameof(ReferenceEquals))
+                                      .OfType<IMethodSymbol>()
```
try to align these with the previous line.

sharwell replied:
➡️ No need to align expressions that the editor doesn't align automatically (which is only the LINQ syntax). If you adjust the indentation, just make it a single indentation, like this:

```csharp
var referenceEqualsMethod = objectType?.GetMembers(nameof(ReferenceEquals))
    .OfType<IMethodSymbol>()
```

CyrusNajmabadi replied:
Note: we do often try to align long chains of linq-as-method calls.  In C# we try to align on the ```.```, and in VB we put the dot on the previous line and align on the name.  It would be nice to preserve that.

That said, going with single indentation is also acceptable.  It's usually strange indentation like this that we try to avoid.

schovanec replied:
I originally had the dots lined up but I think I renamed the variable at some point. I put things back so the dots are lined up.

CyrusNajmabadi commented on src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:
```diff
@@ -88,21 +98,13 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol e
-            var condition = conditionNode as TBinaryExpressionSyntax;
-            if (condition == null)
+            var isEqualityLikeCondition = TryAnalyzeCondition(
+                context, syntaxFacts, referenceEqualsMethod, conditionNode, out var conditionLeft, out var conditionRight, out var isEquals);
```
can you wrap this so the outvars are on the next line.

sharwell commented on src/EditorFeatures/CSharpTest/UseNullPropagation/UseNullPropagationTests.cs:
```diff
@@ -478,5 +478,125 @@ void Main(S? s)
 ");
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseNullPropagation)]
```
:bulb: Also add a `WorkItem` attribute to each of the new tests:

```csharp
[WorkItem("https://github.com/dotnet/roslyn/issues/23043")]
```

zaytsev-victor replied:
I think it should be ```[WorkItem(23043, "https://github.com/dotnet/Roslyn/issues/22699")]``` (:

sharwell replied:
Ugh yes. Except to make me feel a bit better you changed the link by accident 😆 Third try?

```
[WorkItem(23043, "https://github.com/dotnet/roslyn/issues/23043")]
```

sharwell commented on src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:
```diff
@@ -49,18 +50,27 @@ protected AbstractUseNullPropagationDiagnosticAnalyzer()
         }
 
-        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol expressionTypeOpt)
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol expressionTypeOpt, IMethodSymbol referenceEqualsMethod)
```
:bulb: This should be `referenceEqualsMethodOpt`. The `Opt` suffix indicates locally that the argument could be null. The local variables that get passed for this argument should also get renamed.

schovanec replied:
I changed this too. Let me know if I misinterpreted anything here.

sharwell commented on src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:
```diff
@@ -181,9 +183,94 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol e
+
+            var semanticModel = context.SemanticModel;
+            var cancellationToken = context.CancellationToken;
+            var symbol = semanticModel.GetSymbolInfo(invocation, cancellationToken).Symbol;
```
:bulb: `GetSymbolInfo` is quite expensive. Consider updating the code to place the syntax-based null checks (the calls to `IsNullLiteralExpression`) before the call to `GetSymbolInfo`. I'm not overly concerned since you do have the name equality check a few lines above and `ReferenceEquals` isn't a particularly common name, so if it's not easy to make the change feel free to just create a new issue to address this later.

CyrusNajmabadi replied:
Note: this should be fine in practice IMO.  By the time we're checking what symbol this is, we've already ensure the shape is ```ReferenceEquals(e1, e2) ? ... : ...``` that's going to be syntactically rare enough that we'll only being doing the semantic check sparingly.

schovanec replied:
It wasn't to difficult to shuffle this around so I changed it so that `GetSymbolInfo` is called last.

sharwell commented on src/Workspaces/CSharp/Portable/LanguageServices/CSharpSyntaxFactsService.cs:
```diff
@@ -651,6 +651,9 @@ public SyntaxNode GetExpressionOfArgument(SyntaxNode node)
             => (node as ArgumentSyntax).GetRefKind();
 
+        public bool IsArgument(SyntaxNode node)
+            => node is ArgumentSyntax;
```
💭 This makes me wonder why we aren't using `node.IsKind(SyntaxKind.Argument)`, but I see other methods like `IsCastExpression` doing the same thing. Should be fine.

CyrusNajmabadi replied:
It never showed up in perf traces, and it kept things simple i.e. ```IsArgument -> "is Argument"```.  But feel free to change for any nodes that has a 1:1 correspondence with kinds.

schovanec replied:
I removed this method. I don't need it anymore with the change that @CyrusNajmabadi suggested in the code fix class.

schovanec commented:
Thanks for the suggestions @sharwell. I'll make some changes when I get home tonight.

CyrusNajmabadi commented on src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:
```diff
@@ -181,9 +183,94 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol e
         internal static SyntaxNode GetWhenPartMatch(
             ISyntaxFactsService syntaxFacts, ISemanticFactsService semanticFacts, SemanticModel semanticModel, SyntaxNode expressionToMatch, SyntaxNode whenPart)
         {
+            expressionToMatch = GetExpressionIfArgument(syntaxFacts, expressionToMatch);
```
I'm curious why this is needed.  The only time you deal with argumetns is back in TryAnalyzeInvocationCondition.  But at the end of that, you always unwrap arguments to expressions.  So it's unclear how you would get an argument here.

schovanec replied:
The code fix calls into this and passes the node it finds at the location given as part of the diagnostic. With the ReferenceEquals case that node is an argument.

CyrusNajmabadi replied:
Ah, that's a result of an earlier piece of code that needs to change.  Specifically, in the fixer change it from:

```c#
                var conditionalExpression = root.FindNode(diagnostic.AdditionalLocations[0].SourceSpan, getInnermostNodeForTie: true);
                var conditionalPart = root.FindNode(diagnostic.AdditionalLocations[1].SourceSpan);
                var whenPart = root.FindNode(diagnostic.AdditionalLocations[2].SourceSpan);
```

To:

```c#
                var conditionalExpression = root.FindNode(diagnostic.AdditionalLocations[0].SourceSpan, getInnermostNodeForTie: true);
                var conditionalPart = root.FindNode(diagnostic.AdditionalLocations[1].SourceSpan, getInnermostNodeForTie: true);
                var whenPart = root.FindNode(diagnostic.AdditionalLocations[2].SourceSpan, getInnermostNodeForTie: true);
```

Then you should not need to do the argument unwrapping.

CyrusNajmabadi replied:
Sorry i didn't catch that earlier!

schovanec replied:
This change seems to work perfectly, thanks @CyrusNajmabadi.

CyrusNajmabadi commented on src/EditorFeatures/VisualBasicTest/UseNullPropagation/UseNullPropagationTests.vb:
```diff
@@ -321,6 +321,90 @@ Class C
+
+Class C
+    Sub M(o As Object)
+        Dim v = [||]If (ReferenceEquals(o, Nothing), Nothing, o.ToString())
```
so VB has 'ommitted' arguments which are funky and which may end up breaking some of the code htat assumes normal arguments.  Can you try this as a test:

```If (ReferenceEquals(o, ), Nothing, o.ToString())```
```If (ReferenceEquals(, nothing), Nothing, o.ToString())```

Note: i don't expect any actual code change, i just want to make sure we don't crash :)

CyrusNajmabadi commented on src/Features/Core/Portable/UseNullPropagation/AbstractUseNullPropagationDiagnosticAnalyzer.cs:
```diff
@@ -181,9 +183,94 @@ private void AnalyzeSyntax(SyntaxNodeAnalysisContext context, INamedTypeSymbol e
+
+            conditionLeft = syntaxFacts.GetExpressionOfArgument(arguments[0]);
+            conditionRight = syntaxFacts.GetExpressionOfArgument(arguments[1]);
+            return true;
```
i would change this to ```return conditionLeft != null && conditionRight != null; ``` as GetExpressionOfArgument can return null in VB.

chborl commented:
@dotnet-bot retest windows_debug_unit32_prtest please

chborl commented:
@dotnet-bot retest windows_debug_unit64_prtest please

chborl commented:
@dotnet-bot retest windows_release_unit64_prtest please

sharwell commented:
@schovanec I filled out the Ask Mode template in the original post. While reviewing the risk section, I identified two areas for potential improvement:

1. All of the newly added tests focus on one of the expressions being `null` (or `Nothing` in VB). The test suite would be improved if for each one of the existing tests *that you added*, a paired test was added that uses a value other than `null`. These tests ensure that the diagnostic does not flag cases where patterns and/or `ReferenceEquals` are used for a purpose other than a null check.

    ❓ @schovanec Do you want to implement these tests, or do you want me to get them added?

2. A manual differential regression test needs to be run on Roslyn.sln. This is a fairly involved process that I perform myself to ensure each step works as intended. However, since the test strategy has proven enormously valuable to me in other projects in the past, I'll repeat it here should you wish to follow along.
    1. Check out, build, and run the current Roslyn master branch

        <details>

        1. Check out the current **master** branch commit
        1. Set **VisualStudioSetup.Next** as the startup project
        1. Press <kbd>Ctrl</kbd>+<kbd>F5</kbd> to launch an experimental instance

        </details>

    1. Close the copy of Visual Studio where you had Roslyn.sln open
    1. Create a new branch in Git, e.g. **baseline**
    1. Write some code that will trigger the original diagnostic, and apply a Fix All in Solution to correct all instances of it
    1. Commit all changes to the new branch in Git
    1. Close the experimental instance of Visual Studio
    1. Check out the branch for this pull request, and merge the **master** branch into it. Then build an run it like you did for step (i).
    1. Repeat steps (ii) through (vi), but change the branch name to something else, e.g. **modified** (Make sure to create the **modified** branch at the same starting commit where you created **baseline** above)
    1. :arrow_right: Carefully review the difference between the **baseline** and **modified** branches in Git to identify any cases where the code fix did not apply the expected change.

schovanec commented:
Those extra test cases make sense. I can add those but probably not until tomorrow at the earliest. 

> On Dec 7, 2017, at 10:11 AM, Sam Harwell <notifications@github.com> wrote:
> 
> @schovanec I filled out the Ask Mode template in the original post. While reviewing the risk section, I identified two areas for potential improvement:
> 
> All of the newly added tests focus on one of the expressions being null (or Nothing in VB). The test suite would be improved if for each one of the existing tests that you added, a paired test was added that uses a value other than null. These tests ensure that the diagnostic does not flag cases where patterns and/or ReferenceEquals are used for a purpose other than a null check.
> 
> ❓ @schovanec Do you want to implement these tests, or do you want me to get them added?
> 
> A manual differential regression test needs to be run on Roslyn.sln. This is a fairly involved process that I perform myself to ensure each step works as intended. However, since the test strategy has proven enormously valuable to me in other projects in the past, I'll repeat it here should you wish to follow along.
> 
> Check out, build, and run the current Roslyn master branch
> 
> Close the copy of Visual Studio where you had Roslyn.sln open
> 
> Create a new branch in Git, e.g. baseline
> 
> Write some code that will trigger the original diagnostic, and apply a Fix All in Solution to correct all instances of it
> 
> Commit all changes to the new branch in Git
> 
> Close the experimental instance of Visual Studio
> 
> Check out the branch for this pull request, and merge the master branch into it. Then build an run it like you did for step (i).
> 
> Repeat steps (ii) through (vi), but change the branch name to something else, e.g. modified (Make sure to create the modified branch at the same starting commit where you created baseline above)
> 
> ➡️ Carefully review the difference between the baseline and modified branches in Git to identify any cases where the code fix did not apply the expected change.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

schovanec commented:
I added some negative test cases. Let me know if any seem incorrect or if anything obvious (or not obvious) is missing.

I'm going to try and see if I can go through your manual procedure for my own benefit.

Dumb newbie question: If I push my branch after merging from master will it confuse anything in this pull request?

schovanec commented:
@sharwell I was able to run through your regression steps and the result was zero differences between the two branches. I thought this was odd at first since there are new condition patterns that could be recognized. However after some searching there does not appear to be any occurrences of the new condition patterns anywhere in the Roslyn source. There are a few cases that match structurally but are not comparing with null and those were correctly not replaced, which is good.

I look forward to hearing what your results are.

sharwell commented:
@Pilchie for ask mode

Pilchie commented:
Approved - thanks.

jcouv commented:
Nice. Thanks @schovanec!
